### PR TITLE
Documentation for developers to understand continuous build integration

### LINKED ISSUE #896: Documentation for developers to understand continuous build integration
Provide information on how to interpret the build results, including details about artifact names and webhooks. 
Explain the build results and the outcomes of the lsp4ij cron job.

- [ ]  Document the workflow of normal build.yaml execution (PR/Push build execution). - Anusree - In Review
- [ ]  Document how Cron Job builds are executed with respect to build.yaml builds. - Anusree - In Review
- [ ] Document Testing against a Specific Version of LSP4IJ - Anusree - In Review
- [ ]  Document how Slack notifications are triggered based on LSP4IJ PR lists --> Vaisakh - In Review
- [ ]  Document how build results are posted to the Slack channel --> Vaisakh - In Review
- [ ]  Explain how secrets and webhook URLs can be added to the LTI repo/forked repo --> Vaisakh  - In Review
- [ ] Explanation of Slack Messages for Build Results and Listing LSP4IJ PRs --> Vaisakh - In Review
- [ ] Document how to configure and Edit Slack Workflow Builder --> Vaisakh - In Review
- [ ] Document how we can specify Existing LTI Tag or branches to CI/CD builds --> Vaisakh - In Review

### PR DESCRIPTION
**Pull request by anusreelakshmi934**
Fixes #896

### CODE REVIEW

File: docs/user-guide.md [hunk 1]
```diff
@@ -50,6 +51,25 @@ Liberty Tools for IntelliJ IDEA honors the following settings:
 #### Environment Variables:
 - The **JAVA_HOME** system environment variable in the current terminal is used when you run Liberty dev mode in any project.
 
+## Manually install specific release of the LSP4IJ plugin from the Marketplace
+
+When installing the LTI Plugin, the latest version of LSP4IJ from the marketplace will be installed by default.
```
Comment (TrevCraw) [inline docs/user-guide.md:56]:
```suggestion
When installing the Liberty Tools for IntelliJ plugin, the latest version of LSP4IJ from the JetBrains marketplace will be installed by default.
```

File: docs/user-guide.md [hunk 2]
```diff
@@ -50,6 +51,25 @@ Liberty Tools for IntelliJ IDEA honors the following settings:
...
+
+If you prefer to use an older version of LSP4IJ (e.g., the specific version that was tested with the LTI release you are using), you can manually downgrade it by installing the desired version as a dependency from the Marketplace. Link to download a specific LSP4IJ version - https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions
```
Comment (TrevCraw) [inline docs/user-guide.md:58]:
```suggestion
If you prefer to use an older version of LSP4IJ (e.g., the specific version that was tested with the Liberty Tools release you are using), you can manually downgrade it by installing the desired version as a dependency from the JetBrains marketplace. Link to download a specific LSP4IJ version - https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions
```

File: README.md [hunk 3]
```diff
@@ -12,6 +12,10 @@
 
 Liberty Tools for IntelliJ IDEA is an IntelliJ IDEA plugin for developing cloud-native Java applications with [Open Liberty](https://openliberty.io/) and [WebSphere Liberty](https://www.ibm.com/products/websphere-liberty). Iterate fast with Liberty dev mode, code with assistance for MicroProfile and Jakarta EE APIs, and easily edit Liberty configuration files.
 
+Liberty Tools for IntelliJ includes LSP4IJ, an external plugin dependency that is automatically installed with LTI. LSP4IJ is a free and open-source Language Server Protocol (LSP) client developed by Red Hat, enabling seamless language server integration.
+
+Users who wish to change the version of LSP4IJ installed along with LTI can refer to this resource [here](docs/user-guide.md#manually-install-specific-release-of-the-lsp4ij-plugin-from-the-marketplace).
```
Comment (TrevCraw) [inline README.md:17]:
```suggestion
If you wish to change the version of LSP4IJ installed along with Liberty Tools, you can refer to the documentation [here](docs/user-guide.md#manually-install-specific-release-of-the-lsp4ij-plugin-from-the-marketplace).
```

**Review by TrevCraw — CHANGES_REQUESTED** [review]
Overall, looks good! Made some minor suggestions.

I think we should also include a link to the new continuous integration doc from the DEVELOPING.md page.

File: docs/LSP4IJ-Integration.md [hunk 4]
```diff
@@ -0,0 +1,322 @@
+# LSP4IJ Continuous Integration
+
+This guide offers a detailed explanation of the CI setup for LSP4IJ integration, including automated integration and testing with Liberty Tools for IntelliJ. It ensures continuous feedback and provides early warnings when issues arise, allowing us to address them promptly.
+
+- [Workflow of normal build.yaml execution.](#workflow-of-normal-buildyaml-execution)
+    - [Overview](#overview)
```
[the remark this answers is not in the document]
Comment (anusreelakshmi934) [reply docs/LSP4IJ-Integration.md:6]: I initially thought only the main subsections need to be mentioned there. will change that.

File: docs/LSP4IJ-Integration.md [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+1. Configure Pagefile (Windows Only) 
+   - Uses al-cheb/configure-pagefile-action@v1.2 to configure the pagefile size.
+2. Checkout Liberty-Tools-Intellij
+   - Checks out the liberty-tools-intellij repository at the specified LTI Tag/Branch.
+3. Install Required Integration Test Software
+   - Runs a setup script to install necessary software for integration testing.
+   
+_**Steps 4-6 will execute only if useLocalPlugin == true**_
+
```
[the remark this answers is not in the document]
Comment (anusreelakshmi934) [reply docs/LSP4IJ-Integration.md:71]: Ok will change

File: docs/LSP4IJ-Integration.md [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+- `lsp4ijBranch` - Branch or version of LSP4IJ which was used in the build. It is displayed in the artifact name. It can be an LSP4IJ PR number or the value of ‘main’, which means the LSP4IJ ‘main’ branch. The default value is 'default', which means the LSP4IJ used in the build was the latest version in the marketplace.
+
+Below are artifact names resulting from cron job builds. All of the builds ran with the LTI main branch. The first two artifacts come from builds using LSP4IJ PRs. The third artifact comes from a build using the LSP4IJ main branch.
+![Artifacts](images/Artifact-names.png)
+
+### Environment Variable:
+- `SLACK_WEBHOOK_URL`: URL of the Slack webhook for sending notifications.
+
+### Jobs
```
[the remark this answers is not in the document]
Comment (anusreelakshmi934) [reply docs/LSP4IJ-Integration.md:159]: Ok

File: docs/LSP4IJ-Integration.md [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+**Example of a Cron Job Output** - [Here](https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/11011078568The) 
+
+Build run against all 3 OS.
+![Result](images/result-cron-job.png)
+
+Here the output of Job PR details gives the list of PRs in LSP4IJ filtering out drafts and PRs with merge conflicts.It will look like below:
+
+<img alt="Pr-details" height="350" src="images/Pr-details.png" width="400" style="display: block; margin: 0 auto;"/>
+
+The job **Run Lsp4ij Main** is the build which run against the lsp4ij main branch by checking out lsp4ij main branch. The job **Run PR** run by checking out the merge commit SHAs of each PR listed in the job PR details. We can see the PR number against which it is run in the bracket, followed by the merge commit SHA. After all the above jobs are completed, **Run Slack Notification** job runs and we get the Build results as slack notification.
```
[the remark this answers is not in the document]
Comment (anusreelakshmi934) [reply docs/LSP4IJ-Integration.md:212]: PR details is a job and its output is the list of PRs.

[the remark this answers is not in the document]
Comment (anusreelakshmi934) [reply docs/LSP4IJ-Integration.md:212]: @TrevCraw I have modified this section.

File: DEVELOPING.md [hunk 8]
```diff
@@ -61,6 +61,10 @@ To debug the LemMinX Language Server in IntelliJ, complete the following steps.
 
 For more information, see the [LSP console user guide](https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserGuide.md#lsp-console)
 
+### Continuous Integration of LSP4IJ
+
+For details on the Continuous Integration (CI) setup for LSP4IJ integration, refer to the [LSP4IJ Continuous Integration user guide.](docs/LSP4IJ-Continuous-Integration.md)
```
Comment (TrevCraw) [inline DEVELOPING.md:66]:
```suggestion
For details on the Continuous Integration (CI) setup for LSP4IJ integration, refer to the [LSP4IJ Continuous Integration documentation.](docs/LSP4IJ-Continuous-Integration.md) 
```

File: README.md [hunk 9]
```diff
@@ -12,6 +12,10 @@
 
 Liberty Tools for IntelliJ IDEA is an IntelliJ IDEA plugin for developing cloud-native Java applications with [Open Liberty](https://openliberty.io/) and [WebSphere Liberty](https://www.ibm.com/products/websphere-liberty). Iterate fast with Liberty dev mode, code with assistance for MicroProfile and Jakarta EE APIs, and easily edit Liberty configuration files.
 
+Liberty Tools for IntelliJ includes LSP4IJ, an external plugin dependency that is automatically installed with LTI. LSP4IJ is a free and open-source Language Server Protocol (LSP) client developed by Red Hat, enabling seamless language server integration.
```
Comment (TrevCraw) [inline README.md:15]:
```suggestion
Liberty Tools for IntelliJ includes LSP4IJ, an external plugin dependency that is automatically installed with Liberty Tools. LSP4IJ is a free and open-source Language Server Protocol (LSP) client developed by Red Hat, enabling seamless language server integration.
```

File: docs/LSP4IJ-Integration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (TrevCraw) [reply docs/LSP4IJ-Integration.md]: I don't see the change here

File: docs/LSP4IJ-Continuous-Integration.md [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+- Depends on both call-build-workflow-for-each-merge-commit-sha and call-build-workflow-for-lsp4ij-main-branch.
+
+- Always runs regardless of previous job results.
+
+## Results from the Cron Job
+
+**Example of a Cron Job Output** - [Here](https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/11011078568The) 
+
+Each build runs on Linux, Mac, and Windows.
+There are 4 jobs in the build. The first job, **PR details**, outputs a list of PRs in the LSP4IJ repository and displays warnings for any PRs that are either drafts or have merge conflicts.
```
Comment (TrevCraw) [inline docs/LSP4IJ-Continuous-Integration.md:204]:
```suggestion
There are four jobs in the build. The first job, **PR details**, outputs a list of PRs in the LSP4IJ repository and displays warnings for any PRs that are either drafts or have merge conflicts.
```

File: docs/LSP4IJ-Continuous-Integration.md [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+
+<img alt="Pr-details" height="350" src="images/Pr-details.png" width="400" style="display: block; margin: 0 auto;"/>
+
+The next job **Run Lsp4ij Main** is the build which runs against the LSP4IJ main branch by checking out the LSP4IJ main branch. The job **Run PR** runs by checking out the merge commit SHAs of each PR listed in the job **PR details**. We can see the LSP4IJ PR number against which it is run in the bracket, followed by the merge commit SHA. After the above three jobs are completed, **Run Slack Notification** job runs and we get the Build results as slack notification.
```
Comment (TrevCraw) [inline docs/LSP4IJ-Continuous-Integration.md:208]:
```suggestion
The next job, **Run Lsp4ij Main**, is the build which runs against the LSP4IJ main branch by checking out the LSP4IJ main branch. The job **Run PR** runs by checking out the merge commit SHAs of each PR listed in the job **PR details**. The LSP4IJ PR number against which the job is run is displayed in the bracket, followed by the merge commit SHA. After the above three jobs are completed, the **Run Slack Notification** job runs and the Build results are sent as slack notification.
```

File: docs/LSP4IJ-Continuous-Integration.md [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+
+![Result](images/result-cron-job.png)
+
+In the Annotations section, we get the details of the failed tests and also warnings of PRs with merge conflict.
```
Comment (TrevCraw) [inline docs/LSP4IJ-Continuous-Integration.md:212]:
```suggestion
The Annotations section contains errors that show the details of the failed builds and also warnings that show PRs with merge conflicts.
```

File: docs/LSP4IJ-Continuous-Integration.md [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+![Build-failed](images/Build-failed.png)
+
+# Testing against a Specific Version of LSP4IJ
+
+In order to run the LTI build against a particular version of LSP4IJ (stable or nightly), the version can be set in the build.gradle file.
+
+![Version](images/version.png)
+
+To run with a nightly version (For example: 0.4.0-20240813-173701 ), change the line `def lsp4ijVersion = '0.3.0'` to `def lsp4ijVersion = '0.4.0-20240813-173701@nightly'`. Make sure to add **@nightly** to the version number while testing with nightly versions. For stable versions, just specify the version number.
+LSP4IJ nightly builds can be downloaded from the [JetBrains Marketplace](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions/nightly)
```
Comment (TrevCraw) [inline docs/LSP4IJ-Continuous-Integration.md:225]:
```suggestion
LSP4IJ nightly builds can be downloaded from the [JetBrains Marketplace](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions/nightly).
```

File: docs/user-guide.md [hunk 14]
```diff
@@ -50,6 +51,25 @@ Liberty Tools for IntelliJ IDEA honors the following settings:
 #### Environment Variables:
 - The **JAVA_HOME** system environment variable in the current terminal is used when you run Liberty dev mode in any project.
 
+## Manually install specific release of the LSP4IJ plugin from the Marketplace
+
+When installing the Liberty Tools for IntelliJ plugin, the latest version of LSP4IJ from the JetBrains marketplace will be installed by default.
+
+If you prefer to use an older version of LSP4IJ (e.g., the specific version that was tested with the Liberty Tools release you are using), you can manually downgrade it by installing the desired version as a dependency from the JetBrains marketplace. Link to download a specific LSP4IJ version - https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions
```
Comment (TrevCraw) [inline docs/user-guide.md:58]:
```suggestion
If you prefer to use an older version of LSP4IJ (e.g., the specific version that was tested with the Liberty Tools release you are using), you can manually downgrade it by installing the desired version as a dependency from the [JetBrains marketplace](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions).
```

**Review by TrevCraw — CHANGES_REQUESTED** [review]

File: docs/LSP4IJ-Integration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (anusreelakshmi934) [reply docs/LSP4IJ-Integration.md]: Ok Trevor I initally thought the space between those outputs and inputs section. I have modified this section

[the remark this answers is not in the document]
Comment (TrevCraw) [reply docs/LSP4IJ-Integration.md]: looks good now, thank you

**Review by TrevCraw — APPROVED** [review]

**Review by mrglavas — APPROVED** [review]
Reviewed with the team yesterday. Looks good.

**Review by turkeylurkey — APPROVED** [review]

**Pull request merged**